### ISSUE TITLE
Audit linting story

### ISSUE THREAD

**Issue by jasonLaster** [RecordReplay/devtools#6237]
Changes:
- [x] Update our ESLint config
  - [x] #6453
  - [x] #6456
  - [x] #6454
  - [x] #6463
  - [x] #6457
- [x] Update local NPM scripts
  - [x] #6452
  - [x] #6462
- [x] Fix pre-existing lint issues
  - [x] #6458
  - [x] ~~Fix pre-existing lint errors~~
- [ ] Setup Trunk/CI to hold the line going forward
  - [x] #6461

---

We currently have three lint commands

* `npm run lint`
* `next lint`
* `trunk check` 

we run `trunk check` in ci and `next lint` locally. We also effectively disabled `next lint` a month ago when the accordion package was added.

We should
a) standardize on one tool
b) audit the react-hooks warnings

**What is trunk**
Trunk is a new service (https://docs.trunk.io/) that runs all the linters in a performant way. It can also work across languages, so C++, markdown, and perhaps some other compliance checks. The devtools project is in the next ecosystem, so `next lint` would probably be sufficient too

```
info  - Loaded env from /Users/jasonlaster/src/replay/devtools/.env
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

warn  - The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config

./src/devtools/client/debugger/src/components/Editor/Breakpoints/Panel/FirstEditNag.tsx
30:6  Warning: React Hook useEffect has a missing dependency: 'step'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/devtools/client/debugger/src/components/Editor/Breakpoints/Panel/Panel.js
61:6  Warning: React Hook useEffect has missing dependencies: 'dismissNag', 'editor.editor', and 'updateWidth'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/devtools/client/debugger/src/components/Editor/Breakpoints/Panel/Widget.tsx
34:6  Warning: React Hook useEffect has missing dependencies: 'editor.codeMirror', 'insertAt', 'location.line', and 'node'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/devtools/client/debugger/src/components/Editor/LineNumberTooltip.tsx
145:6  Warning: React Hook useEffect has missing dependencies: 'clearHoveredLineNumber', 'editor.codeMirror', and 'setHoveredLineNumber'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/devtools/client/debugger/src/components/Editor/ToggleWidgetButton.tsx
185:6  Warning: React Hook useEffect has a missing dependency: 'editor.codeMirror'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointNavigation.js
46:6  Warning: React Hook useEffect has a missing dependency: 'lastExecutionPoint'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/devtools/client/debugger/src/components/SourceOutline/SourceOutline.tsx
45:6  Warning: React Hook useMemo has an unnecessary dependency: 'listRef'. Either exclude it or remove the dependency array.  react-hooks/exhaustive-deps
53:6  Warning: React Hook useEffect has missing dependencies: 'dispatch' and 'selectedSource'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps
63:6  Warning: React Hook useEffect has a missing dependency: 'outlineSymbols'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps
75:5  Warning: React Hook useCallback has a missing dependency: 'selectLocation'. Either include it or remove the dependency array. If 'selectLocation' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps

./src/devtools/client/inspector/markup/components/MarkupApp.tsx
26:59  Warning: React Hook useEffect has a missing dependency: 'props.inspector'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/devtools/client/inspector/rules/components/RulesApp.tsx
53:5  Warning: React Hook useMemo has missing dependencies: 'onToggleDeclaration', 'onToggleSelectorHighlighter', 'showDeclarationNameEditor', 'showDeclarationValueEditor', 'showNewDeclarationEditor', and 'showSelectorEditor'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/Comments/TranscriptComments/CommentEditor/TipTapEditor.tsx
111:6  Warning: React Hook useEffect has a missing dependency: 'editor'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps
117:6  Warning: React Hook useEffect has a missing dependency: 'editor?.commands'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/Comments/TranscriptComments/CommentSource.tsx
34:6  Warning: React Hook useEffect has missing dependencies: 'comment.sourceLocation', 'createLabels', 'labels.primary', and 'labels.secondary'. Either include them or remove the dependency array. If 'createLabels' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps

./src/ui/components/DevTools.tsx
130:6  Warning: React Hook useEffect has missing dependencies: 'recording?.workspace?.id' and 'userIsAuthor'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps
135:6  Warning: React Hook useEffect has a missing dependency: 'createSession'. Either include it or remove the dependency array. If 'createSession' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps
141:6  Warning: React Hook useEffect has a missing dependency: 'sessionId'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/KeyboardShortcuts.tsx
112:6  Warning: React Hook useEffect has missing dependencies: 'onToggleTheme', 'openFullTextSearch', 'toggleEditFocusMode', 'toggleLeftSidebar', and 'togglePalette'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/Library/Library.tsx
112:6  Warning: React Hook useEffect has missing dependencies: 'currentWorkspaceId', 'setWorkspaceId', 'updateDefaultWorkspace', and 'workspaces'. Either include them or remove the dependency array. If 'setWorkspaceId' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps
124:6  Warning: React Hook useEffect has missing dependencies: 'dismissNag', 'nags', 'pendingWorkspaces?.length', 'setModal', and 'workspaces.length'. Either include them or remove the dependency array. If 'setModal' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps

./src/ui/components/NetworkMonitor/HttpBody.tsx
63:6  Warning: React Hook useMemo has a missing dependency: 'contentType'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/NetworkMonitor/Table.tsx
69:5  Warning: React Hook useMemo has a missing dependency: 'events'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/SkeletonLoader.tsx
45:6  Warning: React Hook useEffect has missing dependencies: 'progress' and 'setFinishedLoading'. Either include them or remove the dependency array. If 'setFinishedLoading' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps

./src/ui/components/Video.tsx
70:6  Warning: React Hook useEffect has missing dependencies: 'currentTime' and 'playback'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/shared/APIKeys.tsx
141:6  Warning: React Hook useEffect has an unnecessary dependency: 'labelRef.current'. Either exclude it or remove the dependency array. Mutable values like 'labelRef.current' aren't valid dependencies because mutating them doesn't re-render the component.  react-hooks/exhaustive-deps

./src/ui/components/shared/CommentTool.tsx
160:6  Warning: React Hook useEffect has missing dependencies: 'addListeners' and 'removeListeners'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/shared/IndexingLoader.tsx
20:6  Warning: React Hook useEffect has a missing dependency: 'progressPercentage'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps
29:6  Warning: React Hook useEffect has a missing dependency: 'isDone'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/shared/NewWorkspaceModal/NewWorkspaceModal.tsx
118:6  Warning: React Hook useEffect has missing dependencies: 'current' and 'setCurrent'. Either include them or remove the dependency array. If 'setCurrent' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps
189:6  Warning: React Hook useEffect has an unnecessary dependency: 'textInputRef.current'. Either exclude it or remove the dependency array. Mutable values like 'textInputRef.current' aren't valid dependencies because mutating them doesn't re-render the component.  react-hooks/exhaustive-deps
272:6  Warning: React Hook useEffect has an unnecessary dependency: 'textInputRef.current'. Either exclude it or remove the dependency array. Mutable values like 'textInputRef.current' aren't valid dependencies because mutating them doesn't re-render the component.  react-hooks/exhaustive-deps

./src/ui/components/shared/Onboarding/DownloadingPage.tsx
12:6  Warning: React Hook useEffect has a missing dependency: 'dismissNag'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/shared/Onboarding/index.tsx
98:6  Warning: React Hook useEffect has missing dependencies: 'current' and 'setCurrent'. Either include them or remove the dependency array. If 'setCurrent' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps

./src/ui/components/shared/OnboardingModal/SingleInviteModal.tsx
32:6  Warning: React Hook useEffect has missing dependencies: 'loading' and 'workspace'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps
55:6  Warning: React Hook useEffect has missing dependencies: 'acceptPendingInvitation' and 'workspace.id'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/shared/PortalDropdown.tsx
25:5  Warning: React Hook useCallback has an unnecessary dependency: 'props.expanded'. Either exclude it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/components/shared/TeamOnboarding.tsx
106:6  Warning: React Hook useEffect has an unnecessary dependency: 'textInputRef.current'. Either exclude it or remove the dependency array. Mutable values like 'textInputRef.current' aren't valid dependencies because mutating them doesn't re-render the component.  react-hooks/exhaustive-deps

./src/ui/components/shared/UserSettingsModal/PreferencesSettings.tsx
79:6  Warning: React Hook useEffect has a missing dependency: 'userSettings.disableLogRocket'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./src/ui/utils/use-key-shortcuts.ts
24:6  Warning: React Hook useEffect has missing dependencies: 'ref' and 'shortcuts'. Either include them or remove the dependency array. Mutable values like 'ref.current' aren't valid dependencies because mutating them doesn't re-render the component.  react-hooks/exhaustive-deps
27:46  Warning: React Hook useEffect has a missing dependency: 'shortcuts'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./pages/login.tsx
13:6  Warning: React Hook useEffect has a missing dependency: 'dispatch'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

./pages/recording/[id]/sourcemap/[sourceId].tsx
102:6  Warning: React Hook useEffect has missing dependencies: 'map', 'source', and 'url'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps
131:6  Warning: React Hook useEffect has missing dependencies: 'recordingId', 'sourceId', and 'store'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./pages/team/invitation/[id].tsx
48:5  Warning: React Hook useEffect has missing dependencies: 'auth0.isAuthenticated', 'claimTeamInvitationCode', 'invitationCode', and 'push'. Either include them or remove the dependency array.  react-hooks/exhaustive-deps

./packages/accordion/index.tsx
105:89  Warning: React Hook useEffect has missing dependencies: 'dispatch' and 'index'. Either include them or remove the dependency array. If 'dispatch' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps

info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
```

**Comment by dapirian** [comment]
Hi there, founder of trunk.io chiming in here! `npm run lint` can run either `next lint` or `trunk check`, so I'll compare those two:

### `next lint`
Runs eslint with the `eslint-plugin-next` plugin - and it creates an eslint config file in your repo root to do that.

### `trunk check` ([docs](https://docs.trunk.io/))
Can easily run the exact same thing as `next lint`, but it _also_ supports:
- autoformatting (`trunk fmt`) dozens of languages/configs
- linting/formatting dozens of other technologies and languages (infra-as-code, ci config files, shell scripts, leaked secrets, etc)
- Preexisting issue detection, so when you turn on new rules you don't have to require fixing all preexisting issues immediately
- Caching
- [VS Code extension (~14k installs)](https://marketplace.visualstudio.com/items?itemName=Trunk.io) to see results inline in your editor
- [GitHub Action](https://github.com/trunk-io/trunk-action) to see results inline in your pull requests
- [Web app](https://app.trunk.io/) to see lint issues over time / audit your codebase

Other info:
- Quick mention for the best way to run prettier (and other autoformatters), which `next lint` doesn't support: [see here](https://github.com/trunk-io/configs#1-turn-off-formatting-errors-use-autoformatters-instead)
- The more checks you can automate, the better - for this repo trunk can run

**Comment by bvaughn** [comment]
I propose swithcing our `npm run lint` script over to run `trunk check` so local and CI setup are at least doing the same thing. (This burned me when I first joined.) –  #6452

Regarding the "quick fix" link–
> 1. Turn off formatting errors, use autoformatters instead

I'd _love_ to do this, but every auto-formatter rule I've tried this morning seems to have a lot of failures on our repo. I don't think people are going to be happy manually fixing things like import order.

**Comment by bvaughn** [comment]
I think the React hooks lint rules will be tricky to update. For example, here's one of the first ones I've looked at:
https://github.com/RecordReplay/devtools/blob/a214920bb5f18016830dbeb83f5734de76f72cb7/src/devtools/client/debugger/src/components/Editor/Breakpoints/Panel/Widget.tsx#L16-L34

There are a lot of missing dependencies for this hook. The naive fix would be to just add them all, but this would likely cause the hook to stop working the way it was intended.

Unfortunately this hook– like most of the others I've looked at– live in parts of the product that I'm not yet familiar with (like new user onboarding) so I'm not even sure what the intended functionality is. 🙂

One strategy would be to tell lint to _ignore_ each of the pre-existing effect rule violations along with a `TODO [authorname]` comment to properly fix the issue. We could follow this up by changing the ESLint config to treat this rule as an _error_ rather than a _warning_ to prevent new violations from being added:
```json
"react-hooks/exhaustive-deps": "error",
```

I think that #6452 takes a bit of the pressure of here either way, since it hides pre-existing warnings from you if you're just using `npm run lint` locally.
